File: baseLib/src/main/java/me/ycdev/android/lib/common/utils/EncodingUtils.java
package me.ycdev.android.lib.common.utils;

import androidx.annotation.NonNull;

@SuppressWarnings({"WeakerAccess", "unused"})
public class EncodingUtils {
    private static final char[] HEX_ARRAY_UPPERCASE = {'0', '1', '2', '3', '4', '5', '6',
            '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'};
    private static final char[] HEX_ARRAY_LOWERCASE = {'0', '1', '2', '3', '4', '5', '6',
            '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};

    /**
     * Encode the data with HEX (Base16) encoding and with uppercase letters.
     */
    public static String encodeWithHex(byte[] bytes) {
        return encodeWithHex(bytes, true);
    }

    public static String encodeWithHex(byte[] bytes, boolean uppercase) {
        if (bytes == null) {
            return "null";
        }
        return encodeWithHex(bytes, 0, bytes.length, uppercase);
    }

    /**
     * Encode the data with HEX (Base16) encoding and with uppercase letters.
     */
    public static String encodeWithHex(@NonNull byte[] bytes, int startPos, int endPos) {
        return encodeWithHex(bytes, startPos, endPos, true);
    }

    public static String encodeWithHex(@NonNull byte[] bytes, int startPos, int endPos, boolean uppercase) {
        if (endPos > bytes.length) {
            endPos = bytes.length;
        }
        final int N = endPos - startPos;
        final char[] HEX_ARRAY = uppercase ? HEX_ARRAY_UPPERCASE : HEX_ARRAY_LOWERCASE;
        char[] hexChars = new char[N * 2];
        for (int i = startPos, j = 0; i < endPos; i++, j += 2) {
            int v = bytes[i] & 0xFF;
            hexChars[j] = HEX_ARRAY[v >>> 4];
            hexChars[j + 1] = HEX_ARRAY[v & 0x0F];
        }
        return new String(hexChars);
    }

    public static byte[] fromHexString(@NonNull String hexStr) {
        hexStr = hexStr.replace(" ", ""); // support spaces
        if (hexStr.length() % 2 != 0) {
            throw new IllegalArgumentException("Bad length: " + hexStr);
        }

        byte[] result = new byte[hexStr.length() / 2];
        for (int i = 0; i < result.length; i++) {
            int high = fromHexChar(hexStr, i * 2) << 4;
            int low = fromHexChar(hexStr, i * 2 + 1);
            result[i] = (byte) ((high | low) & 0xFF);
        }
        return result;
    }

    private static int fromHexChar(String hexStr, int index) {
        char ch = hexStr.charAt(index);
        if (ch >= '0' && ch <= '9') {
            return ch - '0';
        } else if (ch >= 'a' && ch <= 'f') {
            return 10 + (ch - 'a');
        } else if (ch >= 'A' && ch <= 'F') {
            return 10 + (ch - 'A');
        } else {
            throw new IllegalArgumentException("Not hex string: " + hexStr);
        }
    }
}

File: baseLib/src/main/java/me/ycdev/android/lib/common/utils/EncodingUtils.kt
package me.ycdev.android.lib.common.utils

object EncodingUtils {
    private val HEX_ARRAY_UPPERCASE =
        charArrayOf('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F')
    private val HEX_ARRAY_LOWERCASE =
        charArrayOf('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f')

    /**
     * Encode the data with HEX (Base16) encoding
     */
    fun encodeWithHex(bytes: ByteArray?, uppercase: Boolean = true): String {
        return if (bytes == null) {
            "null"
        } else {
            encodeWithHex(bytes, 0, bytes.size, uppercase)
        }
    }

    /**
     * Encode the data with HEX (Base16) encoding
     */
    fun encodeWithHex(
        bytes: ByteArray,
        startPos: Int,
        endPos: Int,
        uppercase: Boolean = true
    ): String {
        var endPosTmp = endPos
        if (endPosTmp > bytes.size) {
            endPosTmp = bytes.size
        }
        val size = endPosTmp - startPos
        val charsArray = if (uppercase) HEX_ARRAY_UPPERCASE else HEX_ARRAY_LOWERCASE
        val hexChars = CharArray(size * 2)
        var i = startPos
        var j = 0
        while (i < endPosTmp) {
            val v = bytes[i].toInt() and 0xFF
            hexChars[j] = charsArray[v.ushr(4)]
            hexChars[j + 1] = charsArray[v and 0x0F]
            i++
            j += 2
        }
        return String(hexChars)
    }

    fun fromHexString(hexStr: String): ByteArray {
        val hexStrTmp = hexStr.replace(" ", "") // support spaces
        if (hexStrTmp.length % 2 != 0) {
            throw IllegalArgumentException("Bad length: $hexStrTmp")
        }

        val result = ByteArray(hexStrTmp.length / 2)
        for (i in result.indices) {
            val high = fromHexChar(hexStrTmp, i * 2) shl 4
            val low = fromHexChar(hexStrTmp, i * 2 + 1)
            result[i] = (high or low and 0xFF).toByte()
        }
        return result
    }

    private fun fromHexChar(hexStr: String, index: Int): Int {
        return when (val ch = hexStr[index]) {
            in '0'..'9' -> ch - '0'
            in 'a'..'f' -> 10 + (ch - 'a')
            in 'A'..'F' -> 10 + (ch - 'A')
            else -> throw IllegalArgumentException("Not hex string: $hexStr")
        }
    }
}
